src/client/executor.rs

```diff
@@ -309,167 +285,152 @@ impl Client {
-                operation_output,
-                connection: conn,
-            }),
-            Err(mut err) => {
-                err.wire_version = conn.stream_description()?.max_wire_version;
+                    err.add_labels_and_update_pin(None, &mut session, None)?;
+                    if err.is_read_retryable() && self.inner.options.retry_writes != Some(false) {
+                        err.add_label(RETRYABLE_WRITE_ERROR);
+                    }
 
-                // Retryable writes are only supported by storage engines with document-level
-                // locking, so users need to disable retryable writes if using mmapv1.
-                if let ErrorKind::Command(ref mut command_error) = *err.kind {
-                    if command_error.code == 20
-                        && command_error.message.starts_with("Transaction numbers")
-                    {
-                        command_error.message = "This MongoDB deployment does not support \
-                                                 retryable writes. Please add retryWrites=false \
-                                                 to your connection string."
-                            .to_string();
+                    let op_retry = match self.get_op_retryability(&op, &session) {
+                        Retryability::Read => err.is_read_retryable(),
+                        Retryability::Write => err.is_write_retryable(),
+                        _ => false,
+                    };
+                    if err.is_pool_cleared() || op_retry {
+                        retry = Some(ExecutionRetry {
+                            prior_txn_number: None,
+                            first_error: err,
+                        });
+                        continue;
+                    } else {
+                        return Err(err);
                     }
                 }
+            };
 
-                self.inner
-                    .topology
-                    .handle_application_error(
-                        server.address.clone(),
-                        err.clone(),
-                        HandshakePhase::after_completion(&conn),
-                    )
-                    .await;
-                // release the connection to be processed by the connection pool
-                drop(conn);
-                // release the selected server to decrement its operation count
-                drop(server);
-
-                if retryability == Retryability::Read && err.is_read_retryable()
-                    || retryability == Retryability::Write && err.is_write_retryable()
-                {
-                    self.execute_retry(&mut op, &mut session, txn_number, err)
-                        .await
-                } else {
-                    Err(err)
-                }
+            if session.is_none() {
+                implicit_session = self.start_implicit_session(&op).await?;
+                session = implicit_session.as_mut();
```

Comment:
IIUC here an implicit session will be started only on the first attempt and reused in subsequent retries. Is this the right behavior? I'm not sure if we should be dropping/re-creating the implicit session after checking out a connection each time around.

Reply:
I believe this is correct - the goal of the upstream change was to minimize the number of sessions created serverside.

Reply:
the retryable writes spec actually _requires_ that the retry attempt use the same session as the initial operation, so this is definitely right, and I assume there are some spec tests that would start failing if we didn't preserve this behavior.

per [spec](https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst#executing-retryable-write-commands):

>  When retrying a write command, drivers MUST resend the command with the same transaction ID.

where "transaction ID" is [defined](https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst#terms) as:

> the top-level lsid and txnNumber fields.

and per the sessions [spec](https://github.com/mongodb/specifications/blob/master/source/sessions/driver-sessions.rst#why-do-automatic-retry-attempts-re-use-a-dirty-implicit-session):

> The retryable writes spec requires that both the original and retry attempt use the same server session. The server will block the retry attempt until the initial attempt completes at which point the retry attempt will continue executing.
For retryable reads that use an implicit session, drivers could choose to use a new server session for the retry attempt however this would lose the information that these two reads are related.